complete reconcilation tests

elevran opened the pull request:
Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

Fixes #2 
Fixes #27

elevran committed:
complete reconcilation tests

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

davidohana commented on pkg/mcshub/controller/controller_test.go:
```diff
@@ -97,411 +82,291 @@ func TestImportGenerated(t *testing.T) {
-	comparePorts(excepted.Spec.Ports, si.Spec.Ports, assertions)
-	compareClusters(excepted.Status.Clusters, si.Status.Clusters, assertions)
-}
+// generate a ServiceExport matching the
```
missing comment

davidohana commented on pkg/mcshub/controller/controller_test.go:
```diff
@@ -97,411 +82,291 @@ func TestImportGenerated(t *testing.T) {
-	err = ser.Client.Get(context.TODO(), req2.NamespacedName, &si)
+	exports := &mcsv1a1.ServiceExportList{}
+	err = ser.Client.List(context.TODO(), exports, client.InNamespace(brokerNS))
+	assert.Nil(err)
```
Should you also validate the import for each export using `validateImportForExportedService()` ?

elevran replied:
I could. I thought it was redundant with the first test.

davidohana replied:
just to ensure no data mix between the two exports / imports

elevran replied:
added

davidohana commented on pkg/mcshub/controller/controller_test.go:
```diff
@@ -97,411 +82,291 @@ func TestImportGenerated(t *testing.T) {
-	if err != nil {
-		t.Error(err)
+	for _, exp := range exports.Items {
+		conflict := lhutil.GetServiceExportCondition(&exp.Status, mcsv1a1.ServiceExportConflict)
```
can refactor this into a function assertNoConflict() as you use it in multiple places.

elevran replied:
refactored into hasConflict(export, conflictStatus) since in some cases expect a conflict

davidohana commented on pkg/mcshub/controller/controller_test.go:
```diff
@@ -97,411 +82,291 @@ func TestImportGenerated(t *testing.T) {
 
-	excepted2 := generateExceptedServiceImport(exp2, es2)
-	compareSi(&excepted2, &si, assertions)
+	assert.Len(exp.GetFinalizers(), 0, "expected no finalizers, found", exp.GetFinalizers())
```
can also be extracted into a func with expected finalizer count parameter

elevran replied:
it would not be any shorter or clearer, IMO.

davidohana commented on pkg/mcshub/controller/controller_test.go:
```diff
@@ -97,411 +82,291 @@ func TestImportGenerated(t *testing.T) {
-func TestUpdateExportWithoutImport(t *testing.T) {
-	exp1, _ := prepareServiceExport(t)
-	preloadedObjects := []runtime.Object{service, exp1}
+// tests that change in non essential export properties don't change the import
```
comment here is not correctly describing the test

elevran replied:
cut and paste from previous. Fixed

davidohana commented on pkg/mcshub/controller/controller_test.go:
```diff
@@ -97,411 +82,291 @@ func TestImportGenerated(t *testing.T) {
+		conflict := lhutil.GetServiceExportCondition(&exp.Status, mcsv1a1.ServiceExportConflict)
+		assert.NotNil(conflict)
+		assert.Equal(exp.GetName() == exp1.GetName(), conflict.Status == corev1.ConditionFalse,
+			"conflict set incorrectly", exp.GetName(), conflict)
```
i think its also important to test conflict resolution when one of the exports is updated or deleted.
also test the conflict status is restored if spoke's export uploader update export (and delete all conditions for example)

elevran replied:
add conflict resolution in latest commit

davidohana reviewed (commented):
done review

elevran committed:
complete unit tests for reconciliation

- add test for conflict resolution
- validate multiple imports when created
- add missing comments

Signed-off-by: Etai Lev Ran <etail@il.ibm.com>

elevran commented:
@davidohana all issues resolved
